File: algorithms-python/list/reverse_list.py
from typing import Optional

from list import util
from list.linked_list import LinkedList, Node


def reverse_list_iteration(head: Node) -> Node:
    if head is None:
        raise ValueError("Head node cannot be null")
    temp: Node = head
    reverse_head: Optional[Node] = None
    while temp is not None:
        next_node: Node = temp.next
        temp.next = reverse_head
        reverse_head = temp
        temp = next_node
    return reverse_head


def reverse_list_recursion(head: Node) -> Node:
    if (head is None) or (head.next is None):
        return head
    reverse_head: Node = reverse_list_iteration(head.next)
    head.next.next = head
    head.next = None
    return reverse_head


def main():
    linked_list: LinkedList = LinkedList()
    linked_list.add_last(Node(1))
    linked_list.add_last(Node(2))
    linked_list.add_last(Node(3))
    linked_list.add_last(Node(4))
    linked_list.add_last(Node(5))
    linked_list.add_last(Node(6))
    linked_list.display()
    # head: Node = reverse_list_iteration(linked_list.head)
    head: Node = reverse_list_recursion(linked_list.head)
    print("\n")
    util.display(head)


if __name__ == '__main__':
    main()

File: algortihms-java/src/com/algo/linkedList/ReverseList.java
package com.algo.linkedList;

public class ReverseList {

	public static void main(String[] args) {
		List list = new List();
		Node n1 = new Node(1);
		Node n2 = new Node(2);
		Node n3 = new Node(3);
		Node n4 = new Node(4);
		Node n5 = new Node(5);
		Node n6 = new Node(6);
		list.add(n1);
		list.add(n2);
		list.add(n3);
		list.add(n4);
		list.add(n5);
		list.add(n6);

		list.display();
		//Node revHead=reverseIteratively(n6);
		Node revHead=reverse(n6);
		System.out.println();
		list.display(revHead);


	}
	
	private static Node reverse(Node head){
		if(head==null||head.next==null){
			return head;
		}
		Node remainingNode=reverse(head.next);
		head.next.next=head;
		head.next=null;
		return remainingNode;
	}
	private static Node reverseIteratively(Node head){
		if(head==null||head.next==null){
			return head;
		}
		Node reverseHead=null;
		while(head!=null){
			Node nextNode=head.next;
			head.next=reverseHead;
			reverseHead=head;
			head=nextNode;
		}
		return reverseHead;
	}
	

}
